<H3>Feature Engineering Assignment on high school students dataset. <br>
Data set has following dimensions and features Students demographics, study habits, parental involvement, extracurricular activities, and academic performance. <br>
The target variable- GPA </H3>

 <h1 style="background-color:#ff6347;">Task 1<h1><br>
 <b>Identify the type of each of the attributes in the dataset</b>


<b>Steps:</b><br>
1. Load the data frame in pandas dataframe
2. View few data records with df.head
3. Describe data set to get initial statistical measures
4. View data values to identify the type of data

In [124]:
# Import required python libraries for processing
import numpy  as np
import pandas as pd


In [125]:
# load the csv file in a pandas data frame using function read_csv
df_org = pd.read_csv('Student_performance_data.csv')
Student_df = df_org
# print first 5 records of dataframe Student_df
print(Student_df.head(5))



   StudentID   Age  Gender  Ethnicity  ParentalEducation  StudyTimeWeekly  \
0       1001  17.0       1          0                  2        19.833723   
1       1002  18.0       0          0                  1        15.408756   
2       1003  15.0       0          2                  3         4.210570   
3       1004  17.0       1          0                  3        10.028829   
4       1005  17.0       1          0                  2         4.672495   

   Absences  Tutoring  ParentalSupport  Extracurricular  Sports  Music  \
0       7.0       1.0              2.0                0     0.0    1.0   
1       0.0       0.0              1.0                0     0.0    0.0   
2      26.0       0.0              2.0                0     0.0    0.0   
3      14.0       0.0              3.0                1     0.0    0.0   
4      17.0       1.0              3.0                0     0.0    0.0   

   Volunteering       GPA  GradeClass  
0             0  2.929196           2  
1           

In [126]:
# print only column heads
Col_list = Student_df.columns
print(Col_list)

Index(['StudentID', 'Age', 'Gender', 'Ethnicity', 'ParentalEducation',
       'StudyTimeWeekly', 'Absences', 'Tutoring', 'ParentalSupport',
       'Extracurricular', 'Sports', 'Music', 'Volunteering', 'GPA',
       'GradeClass'],
      dtype='object')


In [127]:
Student_df.describe()

StudentID          Age       Gender    Ethnicity  ParentalEducation  \
count  2392.000000  2375.000000  2392.000000  2392.000000        2392.000000   
mean   2196.500000    16.468632     0.510870     0.877508           1.746237   
std     690.655244     1.124404     0.499986     1.028476           1.000411   
min    1001.000000    15.000000     0.000000     0.000000           0.000000   
25%    1598.750000    15.000000     0.000000     0.000000           1.000000   
50%    2196.500000    16.000000     1.000000     0.000000           2.000000   
75%    2794.250000    17.000000     1.000000     2.000000           2.000000   
max    3392.000000    18.000000     1.000000     3.000000           4.000000   

       StudyTimeWeekly     Absences     Tutoring  ParentalSupport  \
count      2358.000000  2380.000000  2367.000000      2378.000000   
mean          9.761567    14.546639     0.301225         2.122372   
std           5.658241     8.463651     0.458887         1.122941   
min           0.001057     0.000000     0.000000         0.000000   
25%           5.030526     7.000000     0.000000         1.000000   
50%           9.653180    15.000000     0.000000         2.000000   
75%          14.408107    22.000000     1.000000         3.000000   
max          19.978094    29.000000     1.000000         4.000000   

       Extracurricular       Sports        Music  Volunteering          GPA  \
count      2392.000000  2381.000000  2180.000000   2392.000000  2392.000000   
mean          0.383361     0.303234     0.197248      0.157191     1.906186   
std           0.486307     0.459752     0.398012      0.364057     0.915156   
min           0.000000     0.000000     0.000000      0.000000     0.000000   
25%           0.000000     0.000000     0.000000      0.000000     1.174803   
50%           0.000000     0.000000     0.000000      0.000000     1.893393   
75%           1.000000     1.000000     0.000000      0.000000     2.622216   
max           1.000000     1.000000     1.000000      1.000000     4.000000   

        GradeClass  
count  2392.000000  
mean      2.983696  
std       1.233908  
min       0.000000  
25%       2.000000  
50%       4.000000  
75%       4.000000  
max       4.000000

In quantative feature only GPA,Study TimeWeekly is continous other are discrete wholenumbers without fractions.
The above function gives good insights about spread of Quantitative features 

In [128]:
#encoding  logic for column name and types of data
task1_df = pd.DataFrame()
task1_df['Columns']=Col_list
Nominal ,Ratio,Ordinal,Interval,C,D= 'Nominal','Ratio','Ordinal','Interval','Continous','Discrete'
task1_df['Type'] = [Nominal,Ratio,Nominal,Nominal,Nominal,Ratio,Ratio,Nominal,Nominal,Nominal,Nominal,Nominal,Nominal,Ratio,Ordinal]
reason_tmp=[]
task1_df['Binary']=[False,False,True,False,False,False,False,True,False,True,True,True,True,False,False]
task1_df['Discrete_continous']=[D,D,D,D,D,C,D,D,D,D,D,D,D,C,D]
for type,b in zip(task1_df['Type'],task1_df['Binary']) :
    if type in [Nominal,Ordinal]:
        Prefix ='*Catergorical,Qualitative*' 
        if type == Nominal:
            order ="*Discrete labels not ordered or ranked*"
        else:
             order ="*Discrete labels with order, but no  same difference between order* "

    else:
        Prefix ='*Numerical,Quantative*'
        if type == Ratio:
            order ="*ordered units that have the same difference and zero reference.* "
        else:
             order ="*ordered units that have the same difference and No zero reference. *"
    if b=='True':
        Binary =" *special features takes on binary values*"
    else:
        Binary =' '
    reason_tmp.append(Prefix + ' '+order + Binary)
         
task1_df['Reason'] =reason_tmp

markdown_table = task1_df.to_markdown(index=False)
print(markdown_table)

| Columns           | Type    | Binary   | Discrete_continous   | Reason                                                                                         |
|:------------------|:--------|:---------|:---------------------|:-----------------------------------------------------------------------------------------------|
| StudentID         | Nominal | False    | Discrete             | *Catergorical,Qualitative* *Discrete labels not ordered or ranked*                             |
| Age               | Ratio   | False    | Discrete             | *Numerical,Quantative* *ordered units that have the same difference and zero reference.*       |
| Gender            | Nominal | True     | Discrete             | *Catergorical,Qualitative* *Discrete labels not ordered or ranked*                             |
| Ethnicity         | Nominal | False    | Discrete             | *Catergorical,Qualitative* *Discrete labels not ordered or ranked*                             |
| ParentalEducation | 

 <h1 style="background-color:#ff6347;">Task 2<h1><br>
 <b>Name the attribute(s) which have missing data. How many data points are missing? How will you handle missing data ?</b>


In [129]:
Student_df.shape

(2392, 15)

There are 2392 rows and 15 columns

In [130]:
Student_df.count()

StudentID            2392
Age                  2375
Gender               2392
Ethnicity            2392
ParentalEducation    2392
StudyTimeWeekly      2358
Absences             2380
Tutoring             2367
ParentalSupport      2378
Extracurricular      2392
Sports               2381
Music                2180
Volunteering         2392
GPA                  2392
GradeClass           2392
dtype: int64

We see that there are  2392 rows and columns like <b>Age,StudyTimeWeekly,Absences,Tutoring,ParentalSupport,Sports,Music </b>have missing value

In [131]:
Student_df.isna().sum()

StudentID              0
Age                   17
Gender                 0
Ethnicity              0
ParentalEducation      0
StudyTimeWeekly       34
Absences              12
Tutoring              25
ParentalSupport       14
Extracurricular        0
Sports                11
Music                212
Volunteering           0
GPA                    0
GradeClass             0
dtype: int64

The above table shows columns,missing values counts.

The missing values will be handled with technniques like
1. Dropping rows with null values
2. imputing rows with Zero,mean,median,mode,ramdom values.

 <h1 style="background-color:#ff6347;">Task 3<h1><br>
 <b>Identify the type of each of the attributes in the dataset</b>


we will not use the strategy of dropping rows as there will be loss of training and testing data volume.<br>
We will use different imutation technnique based on data type and domain context.

Age has 17 missing values.Data is missing completely at random.
Range of age 15 years to 18.
it is a whole number (discrete)hence will be imputing with the mode value

In [132]:
temp = Student_df['Age'].mode()
Student_df["Age"].fillna(temp.values[0],inplace=True) # Replace missing values with the mode of the column
#Student_df["Age"].count()

StudyTimeWeekly has 34 missing values.Data is missing completely at random.
Range of StudyTimeWeekly is  >0 & <20
it is a continous value  hence will be imputing with the mean value

In [133]:
temp = Student_df['StudyTimeWeekly'].mean()
Student_df["StudyTimeWeekly"].fillna(temp,inplace=True) # Replace missing values with the mode of the column
#Student_df["StudyTimeWeekly"].count()

Absences has 12 missing values.Data is missing completely at random.
Range of StudyTimeWeekly is  0 to 29
it is a whole number(discrete value)  hence will be imputing with the mode value

In [134]:
temp = Student_df['Absences'].mode()
Student_df["Absences"].fillna(temp.values[0],inplace=True) # Replace missing values with the mode of the column
#Student_df["Absences"].count()

Tutoring has 25 missing values.Data is missing completely at random.
Range of Tutoring is  0 or 1
it is a categorical hence will be imputing with the mode value


In [135]:
temp = Student_df['Tutoring'].mode()
Student_df["Tutoring"].fillna(temp.values[0],inplace=True) # Replace missing values with the mode of the column
#Student_df["Tutoring"].count()

ParentalSupport has 14 missing values.Data is missing completely at random.
Range of Tutoring is  0 to 4
it is a categorical hence will be imputing with the mode value


In [136]:
temp = Student_df['ParentalSupport'].mode()
Student_df["ParentalSupport"].fillna(temp.values[0],inplace=True) # Replace missing values with the mode of the column
#Student_df["ParentalSupport"].count()

Sports has 11 missing values.Data is missing completely at random.
Range of Tutoring is  0 or 1
it is a categorical hence will be imputing with the mode value


In [137]:
temp = Student_df['Sports'].mode()
Student_df["Sports"].fillna(temp.values[0],inplace=True) # Replace missing values with the mode of the column
#Student_df["Sports"].count()

Music has 212 missing values.Data is missing completely at random.
Range of Tutoring is  0 or 1
it is a categorical hence will be imputing with the mode value


In [138]:
temp = Student_df['Music'].mode()
Student_df["Music"].fillna(temp.values[0],inplace=True) # Replace missing values with the mode of the column
#Student_df["Music"].count()


In [139]:
print(Student_df.count())

print(Student_df.isna().sum())

StudentID            2392
Age                  2392
Gender               2392
Ethnicity            2392
ParentalEducation    2392
StudyTimeWeekly      2392
Absences             2392
Tutoring             2392
ParentalSupport      2392
Extracurricular      2392
Sports               2392
Music                2392
Volunteering         2392
GPA                  2392
GradeClass           2392
dtype: int64
StudentID            0
Age                  0
Gender               0
Ethnicity            0
ParentalEducation    0
StudyTimeWeekly      0
Absences             0
Tutoring             0
ParentalSupport      0
Extracurricular      0
Sports               0
Music                0
Volunteering         0
GPA                  0
GradeClass           0
dtype: int64


 <h1 style="background-color:#ff6347;">Task 4<h1><br>
 <b>Perform normalization of all numeric attributes.</b>

The following are features have <b>number</b> values.['StudentID', 'Age', 'Gender', 'Ethnicity', 'ParentalEducation',
       'StudyTimeWeekly', 'Absences', 'Tutoring', 'ParentalSupport',
       'Extracurricular', 'Sports', 'Music', 'Volunteering', 'GPA',
       'GradeClass'] <br>
The dataset has all numberical values.<BR>
['StudentID'] is nominal unique identifier of data records and so not used in any model prediction.<br>
['Gender','Tutoring', 'Extracurricular', 'Sports', 'Music', 'Volunteering',] are binary nominal categories so no normalization is required. <br>
[ 'Ethnicity', 'ParentalEducation', 'ParentalSupport'] are categorical value and weuse <b>one hot encoding</b> to  normalization  them. and remove implicit order<br>
Feature normalization is mostly needed to eliminate the effect of several quantitative features measured on different scales.<br>
['Age','StudyTimeWeekly'] can only be normalised.we  will use <b>min max scaling to preserve the distribution</b><br>
MinMaxScaler doesn’t reduce the effect of outliers. but our data set feature ranges show that there are no outliers.so we can use minmax scaling<br>

['GPA'] is target variable .A target variable with a large spread of values, in turn, may result in large error gradient values causing weight values to change dramatically, making the learning process unstable so requires scaling.but since GPA spread is less we neednot normalise.<br>
['GradeClass'] is a linearly derived variable from target variable GPA.Normalisation of target variable is not required, based on above explanation<br>

In [140]:
from sklearn.preprocessing import MinMaxScaler
# scaling age 
scaler = MinMaxScaler().fit(pd.DataFrame(Student_df["Age"]))
Student_df["Age"] =scaler.transform(pd.DataFrame(Student_df["Age"]))

# scaling StudyTimeWeekly 
scaler = MinMaxScaler().fit(pd.DataFrame(Student_df["StudyTimeWeekly"]))
Student_df["StudyTimeWeekly"] =scaler.transform(pd.DataFrame(Student_df["StudyTimeWeekly"]))
Student_df[["Age",'StudyTimeWeekly']]
Student_df.head(5)

StudentID       Age  Gender  Ethnicity  ParentalEducation  StudyTimeWeekly  \
0       1001  0.666667       1          0                  2         0.992773   
1       1002  1.000000       0          0                  1         0.771270   
2       1003  0.000000       0          2                  3         0.210718   
3       1004  0.666667       1          0                  3         0.501965   
4       1005  0.666667       1          0                  2         0.233840   

   Absences  Tutoring  ParentalSupport  Extracurricular  Sports  Music  \
0       7.0       1.0              2.0                0     0.0    1.0   
1       0.0       0.0              1.0                0     0.0    0.0   
2      26.0       0.0              2.0                0     0.0    0.0   
3      14.0       0.0              3.0                1     0.0    0.0   
4      17.0       1.0              3.0                0     0.0    0.0   

   Volunteering       GPA  GradeClass  
0             0  2.929196           2  
1             0  3.042915           1  
2             0  0.112602           4  
3             0  2.054218           3  
4             0  1.288061           4

In [141]:
Student_df = Student_df.astype({'Ethnicity':int, 'ParentalEducation':int, 'ParentalSupport':int})

In [142]:
from sklearn.preprocessing import OneHotEncoder

print(Student_df["Ethnicity"].unique())

ohe = OneHotEncoder(handle_unknown='ignore',sparse_output=False).set_output(transform='pandas')
ohe_done = ohe.fit_transform(Student_df[["Ethnicity"]])
Student_df = pd.concat([Student_df,ohe_done],axis =1)
ohe_done = ohe.fit_transform(Student_df[["ParentalEducation"]])
Student_df =  pd.concat([Student_df,ohe_done],axis =1)
ohe_done = ohe.fit_transform(Student_df[["ParentalSupport"]])
Student_df =  pd.concat([Student_df,ohe_done],axis =1).drop(columns=['Ethnicity','ParentalEducation','ParentalSupport'])
print(Student_df.head())

[0 2 1 3]
   StudentID       Age  Gender  StudyTimeWeekly  Absences  Tutoring  \
0       1001  0.666667       1         0.992773       7.0       1.0   
1       1002  1.000000       0         0.771270       0.0       0.0   
2       1003  0.000000       0         0.210718      26.0       0.0   
3       1004  0.666667       1         0.501965      14.0       0.0   
4       1005  0.666667       1         0.233840      17.0       1.0   

   Extracurricular  Sports  Music  Volunteering  ...  ParentalEducation_0  \
0                0     0.0    1.0             0  ...                  0.0   
1                0     0.0    0.0             0  ...                  0.0   
2                0     0.0    0.0             0  ...                  0.0   
3                1     0.0    0.0             0  ...                  0.0   
4                0     0.0    0.0             0  ...                  0.0   

   ParentalEducation_1  ParentalEducation_2  ParentalEducation_3  \
0                  0.0          

In [143]:
print(Student_df.columns)

Index(['StudentID', 'Age', 'Gender', 'StudyTimeWeekly', 'Absences', 'Tutoring',
       'Extracurricular', 'Sports', 'Music', 'Volunteering', 'GPA',
       'GradeClass', 'Ethnicity_0', 'Ethnicity_1', 'Ethnicity_2',
       'Ethnicity_3', 'ParentalEducation_0', 'ParentalEducation_1',
       'ParentalEducation_2', 'ParentalEducation_3', 'ParentalEducation_4',
       'ParentalSupport_0', 'ParentalSupport_1', 'ParentalSupport_2',
       'ParentalSupport_3', 'ParentalSupport_4'],
      dtype='object')
